File: src/main/java/ltdjms/discord/aiagent/commands/ToolExecutionListener.java
package ltdjms.discord.aiagent.commands;

import java.util.List;
import java.util.function.Consumer;
import javax.inject.Inject;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import ltdjms.discord.aichat.services.MessageSplitter;
import ltdjms.discord.shared.di.JDAProvider;
import ltdjms.discord.shared.events.DomainEvent;
import ltdjms.discord.shared.events.LangChain4jToolExecutedEvent;
import net.dv8tion.jda.api.entities.Guild;
import net.dv8tion.jda.api.entities.channel.middleman.GuildChannel;
import net.dv8tion.jda.api.entities.channel.middleman.MessageChannel;

/**
 * LangChain4J 工具執行事件監聽器。
 *
 * <p>接收 {@link LangChain4jToolExecutedEvent} 並將通知訊息發送到對應的 Discord 討論串。
 */
public final class ToolExecutionListener implements Consumer<DomainEvent> {

  private static final Logger LOGGER = LoggerFactory.getLogger(ToolExecutionListener.class);

  @Inject
  public ToolExecutionListener() {}

  @Override
  public void accept(DomainEvent event) {
    if (event instanceof LangChain4jToolExecutedEvent e) {
      handleToolExecuted(e);
    }
  }

  private void handleToolExecuted(LangChain4jToolExecutedEvent event) {
    try {
      MessageChannel channel = resolveMessageChannel(event.guildId(), event.channelId());
      if (channel == null) {
        LOGGER.warn("無法取得工具通知頻道: guildId={}, channelId={}", event.guildId(), event.channelId());
        return;
      }

      String message =
          event.success()
              ? "✅ 工具「%s」執行成功".formatted(event.toolName())
              : "❌ 工具「%s」執行失敗：%s".formatted(event.toolName(), event.result());

      List<String> chunks = MessageSplitter.split(message);
      for (String chunk : chunks) {
        channel.sendMessage(chunk).queue();
      }

    } catch (Exception e) {
      LOGGER.error("處理工具執行事件時發生錯誤", e);
    }
  }

  private MessageChannel resolveMessageChannel(long guildId, long channelId) {
    Guild guild = JDAProvider.getJda().getGuildById(guildId);
    if (guild == null) {
      return null;
    }

    GuildChannel guildChannel = guild.getGuildChannelById(channelId);
    if (guildChannel instanceof MessageChannel messageChannel) {
      return messageChannel;
    }

    return guild.getThreadChannelById(channelId);
  }
}

File: packages/ai/src/listeners/tool-execution-listener.ts
import type { DomainEvent, DiscordRuntimeGateway } from '@ltdjms/shared';
import type {
  LangChain4jToolExecutedEvent,
  LangChain4jToolExecutionStartedEvent,
} from '@ltdjms/shared';
import type { GuildTextBasedChannel, Message } from 'discord.js';
import { MessageSplitter } from '../services/MessageSplitter.js';
import pino from 'pino';

/**
 * Sends Discord notifications when LangChain agent tools start/complete.
 * Matches Java ToolExecutionListener.
 */
export class ToolExecutionListener {
  private readonly splitter = new MessageSplitter();
  private readonly logger: pino.Logger;

  constructor(
    private readonly runtimeGateway: DiscordRuntimeGateway,
    logger?: pino.Logger,
  ) {
    this.logger = logger ?? pino({ name: 'tool-execution-listener' });
  }

  accept(event: DomainEvent | null | undefined): void {
    if (!event) {
      return;
    }

    if (event.eventType === 'langchain4j_tool_execution_started') {
      this.handleToolExecutionStarted(event as LangChain4jToolExecutionStartedEvent);
      return;
    }

    if (event.eventType === 'langchain4j_tool_executed') {
      this.handleToolExecuted(event as LangChain4jToolExecutedEvent);
    }
  }

  private handleToolExecutionStarted(event: LangChain4jToolExecutionStartedEvent): void {
    try {
      const channel = this.resolveMessageChannel(event.guildId, event.channelId);
      if (!channel) {
        this.logger.warn(
          { guildId: event.guildId, channelId: event.channelId },
          '無法取得工具通知頻道',
        );
        return;
      }

      const message = `🤖 我先執行這一步：正在呼叫工具「${event.toolName}」...`;
      void this.sendChunks(channel, message);
    } catch (error) {
      this.logger.error({ err: error }, '處理工具開始事件時發生錯誤');
    }
  }

  private handleToolExecuted(event: LangChain4jToolExecutedEvent): void {
    try {
      const channel = this.resolveMessageChannel(event.guildId, event.channelId);
      if (!channel) {
        this.logger.warn(
          { guildId: event.guildId, channelId: event.channelId },
          '無法取得工具通知頻道',
        );
        return;
      }

      const message = event.success
        ? `✅ 工具「${event.toolName}」執行成功`
        : `❌ 工具「${event.toolName}」執行失敗：${event.result}`;

      void this.sendChunks(channel, message);
    } catch (error) {
      this.logger.error({ err: error }, '處理工具執行事件時發生錯誤');
    }
  }

  private resolveMessageChannel(
    guildId: string,
    channelId: string,
  ): GuildTextBasedChannel | null {
    const channel =
      this.runtimeGateway.findGuildChannel(guildId, channelId) ??
      this.runtimeGateway.findThreadChannel(guildId, channelId);

    if (!channel || typeof channel !== 'object') {
      return null;
    }

    const candidate = channel as { isTextBased?: () => boolean; send?: unknown };
    if (typeof candidate.isTextBased === 'function' && candidate.isTextBased()) {
      return channel as GuildTextBasedChannel;
    }

    return null;
  }

  private async sendChunks(channel: GuildTextBasedChannel, message: string): Promise<void> {
    for (const chunk of this.splitter.split(message)) {
      if (chunk.trim()) {
        await channel.send(chunk).catch(() => undefined);
      }
    }
  }
}
